b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,395 @@
+    delete this._connections.get(connection.number);
+  },
+};
+
+// Represents a connection to the server
+function Connection(input, output, server, port, outgoingPort, number) {
+  DEBUG && debug("Open a new connection " + number + " on port " + outgoingPort);
+
+  // Stream of incoming data
+  this.input = input;
+
+  // Stream for outgoing data
+  this.output = output;
+
+  // Server associated with this connection
+  this.server = server;
+
+  // Port on which the server is running
+  this.port = port;
+
+  // Outgoing poort used by this connection
+  this._outgoingPort = outgoingPort;
+
+  // The serial number of this connection
+  this.number = number;
+
+  // This allows a connection to disambiguate between a peer initiating a
+  // close and the socket being forced closed on shutdown.
+  this._closed = false;
+}
+Connection.prototype = {
+  // Closes this connection's input/output streams
+  close: function() {
+    if (this._closed) {
+      return;
+    }
+
+    DEBUG && debug("Close connection " + this.number + " on port " + this._outgoingPort);
+
+    this.input.close();
+    this.output.close();
+    this._closed = true;
+
+    let server = this.server;
+    server._connectionClosed(this);
+
+    // If an error triggered a server shutdown, act on it now
+    if (server._doQuit) {
+      server.stop();
+    }
+  },
+};
+
+function streamClosed(e) {
+  return e === Cr.NS_BASE_STREAM_CLOSED ||
+         (typeof e === "object" && e.result === Cr.NS_BASE_STREAM_CLOSED);
+}
+
+// Parse and dispatch incoming events from client
+function EventHandler(connection) {
+  this._connection = connection;
+
+  this._output = Cc["@mozilla.org/intl/converter-output-stream;1"]
+                   .createInstance(Ci.nsIConverterOutputStream);
+  this._output.init(this._connection.output, "UTF-8", 0, 0x0000);
+
+  this._input = Cc["@mozilla.org/intl/converter-input-stream;1"]
+                  .createInstance(Ci.nsIConverterInputStream);
+  this._input.init(this._connection.input, "UTF-8", 0,
+                   Ci.nsIConverterInputStream.DEFAULT_REPLACEMENT_CHARACTER);
+}
+EventHandler.prototype = {
+  // nsIInputStreamCallback
+  onInputStreamReady: function(input) {
+    DEBUG && debug("onInputStreamReady(input=" + input + ") on thread " +
+                   Services.tm.currentThread + " (main is " +
+                   Services.tm.mainThread + ")");
+
+    try {
+      let available = 0, numChars = 0, fullMessage = "";
+
+      // Read and concat messages from input stream buffer
+      do {
+        let partialMessage = {};
+
+        available = input.available();
+        numChars = this._input.readString(available, partialMessage);
+
+        fullMessage += partialMessage.value;
+      } while(numChars < available);
+
+      if (numChars != 0 && fullMessage.length > 0) { // While last readString contains something
+        // Handle concatenated JSON string
+        fullMessage = '[' + fullMessage.replace(/}{/g, '},{') + ']';
```

Comment:
We can move this line to some function like `_sanitizeMessage` and add more JSON format check, e.g.
- if the string doesn't start with '{', discard the part before the first found of '}{'.
- if the string doesn't end with '}', discard the part after the last found of '}{'